src/finch/compile/bufferized_ndarray.py

```diff
@@ -7,46 +9,99 @@
 from .. import finch_notation as ntn
 from ..algebra import Tensor
 from ..codegen import NumpyBuffer
-from ..finch_assembly import AssemblyStructFType
+from ..finch_assembly import AssemblyStructFType, TupleFType, Variable
 from ..symbolic import ftype
 from . import looplets as lplt
 from .lower import FinchTensorFType
 
 
+class StrideState(Enum):
+    FREE = 0
+    ACTIVE = 1
+    VISITED = 2
+
+
+class BufferizedNDArrayStrideState:
+    vars: tuple[Variable, ...]
+    strides: tuple[StrideState, ...]
+
+    def __init__(
+        self, vars: tuple[Variable, ...], strides: tuple[StrideState, ...]
+    ) -> None:
+        self.vars = vars
+        self.strides = strides
+
+    def __str__(self) -> str:
+        return f"StrideState({self.vars}, {self.strides})"
+
+    def __hash__(self) -> int:
+        return hash((self.vars, self.strides))
+
+    @classmethod
+    def full(
+        cls, vars: tuple[Variable, ...], fv: StrideState = StrideState.FREE
+    ) -> "BufferizedNDArrayStrideState":
+        return BufferizedNDArrayStrideState(vars, tuple(fv for _ in range(len(vars))))
+
+    def ndim(self, skip_visited: bool = True) -> int:
+        return len(
+            [s for s in self.strides if not skip_visited or s != StrideState.VISITED]
+        )
+
+    def get_shape(self, shape: Iterable[Any]) -> tuple[Any, ...]:
+        return tuple(
+            elem
+            for elem, stride in zip(shape, self.strides, strict=True)
+            if stride != StrideState.VISITED
+        )
+
+    def get_active_stride(self) -> int:
+        return self.strides.index(StrideState.ACTIVE)
+
+    def activate_step(self, var: Variable) -> "BufferizedNDArrayStrideState":
+        assert self.strides.count(StrideState.ACTIVE) == 0
+        idx = self.vars.index(var)
+        assert self.strides[idx] == StrideState.FREE
+        result = list(self.strides)
+        result[idx] = StrideState.ACTIVE
+        return BufferizedNDArrayStrideState(self.vars, tuple(result))
+
+    def visit_step(self) -> "BufferizedNDArrayStrideState":
+        assert self.strides.count(StrideState.ACTIVE) == 1
+        return BufferizedNDArrayStrideState(
+            self.vars,
+            tuple(
+                s if s == StrideState.FREE else StrideState.VISITED
+                for s in self.strides
+            ),
+        )
+
+
 class BufferizedNDArray(Tensor):
     def __init__(self, arr: np.ndarray):
         itemsize = arr.dtype.itemsize
         for stride in arr.strides:
             if stride % itemsize != 0:
                 raise ValueError("Array must be aligned to multiple of itemsize")
-        self.strides = tuple(stride // itemsize for stride in arr.strides)
```

Comment:
Current strides logic and `as_strided` usage breaks NumPy causing incorrect outputs. I simplified it a bit.